File: mobile/android/base/java/org/mozilla/gecko/drawable/ShiftDrawable.java
package org.mozilla.gecko.drawable;

/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.graphics.Canvas;
import android.graphics.Path;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.animation.Interpolator;
import android.view.animation.LinearInterpolator;

/**
 * A drawable to keep shifting its wrapped drawable.
 * Assume the wrapped drawable value is "00000010", this class will keep drawing in this way
 * <p>
 * 00000010 -> 00000001 -> 10000000 -> 01000000 -> 00100000 -> ...
 * <p>
 * This drawable will keep drawing until be invisible.
 */
public class ShiftDrawable extends DrawableWrapper {

    /**
     * An animator to trigger redraw and update offset-of-shifting
     */
    private final ValueAnimator mAnimator = ValueAnimator.ofFloat(0f, 1f);

    /**
     * Visible rectangle, wrapped-drawable is resized and draw in this rectangle
     */
    private final Rect mVisibleRect = new Rect();

    /**
     * Canvas will clip itself by this Path. Used to draw rounded head.
     */
    private final Path mPath = new Path();

    // align to ScaleDrawable implementation
    private static final int MAX_LEVEL = 10000;

    private static final int DEFAULT_DURATION = 1000;

    public ShiftDrawable(@NonNull Drawable d) {
        this(d, DEFAULT_DURATION);
    }

    public ShiftDrawable(@NonNull Drawable d, int duration) {
        this(d, duration, new LinearInterpolator());
    }

    public ShiftDrawable(@NonNull Drawable d, int duration, @Nullable Interpolator interpolator) {
        super(d);
        mAnimator.setDuration(duration);
        mAnimator.setRepeatCount(ValueAnimator.INFINITE);
        mAnimator.setInterpolator((interpolator == null) ? new LinearInterpolator() : interpolator);
        mAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                if (isVisible()) {
                    invalidateSelf();
                }
            }
        });
        mAnimator.start();
    }

    public Animator getAnimator() {
        return mAnimator;
    }

    /**
     * {@inheritDoc}
     * <p>
     * override to enable / disable animator as well.
     */
    @Override
    public boolean setVisible(final boolean visible, final boolean restart) {
        final boolean result = super.setVisible(visible, restart);
        if (isVisible()) {
            mAnimator.start();
        } else {
            mAnimator.end();
        }
        return result;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected void onBoundsChange(Rect bounds) {
        super.onBoundsChange(bounds);
        updateBounds();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected boolean onLevelChange(int level) {
        final boolean result = super.onLevelChange(level);
        updateBounds();
        return result;
    }

    @Override
    public void draw(@NonNull Canvas canvas) {
        final Drawable wrapped = getWrappedDrawable();
        final float fraction = mAnimator.getAnimatedFraction();
        final int width = mVisibleRect.width();
        final int offset = (int) (width * fraction);

        final int stack = canvas.save();

        // To apply mPath, then we have rounded-head
        canvas.clipPath(mPath);

        // To draw left-half part of Drawable, shift from right to left
        canvas.save();
        canvas.translate(-offset, 0);
        wrapped.draw(canvas);
        canvas.restore();

        // Then to draw right-half part of Drawable
        canvas.save();
        canvas.translate(width - offset, 0);
        wrapped.draw(canvas);
        canvas.restore();

        canvas.restoreToCount(stack);
    }

    private void updateBounds() {
        final Rect b = getBounds();
        final int width = (int) ((float) b.width() * getLevel() / MAX_LEVEL);
        mVisibleRect.set(b.left, b.top, b.left + width, b.height());

        // to create path to help drawing rounded head. mPath is enclosed by mVisibleRect
        final float radius = b.height() / 2;
        mPath.reset();

        // The added rectangle width is smaller than mVisibleRect, due to semi-circular.
        mPath.addRect(mVisibleRect.left,
                mVisibleRect.top, mVisibleRect.right - radius,
                mVisibleRect.height(),
                Path.Direction.CCW);
        // To add semi-circular
        mPath.addCircle(mVisibleRect.right - radius, radius, radius, Path.Direction.CCW);
    }
}

File: mobile/android/android-components/components/ui/progress/src/main/java/mozilla/components/ui/progress/ShiftDrawable.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package mozilla.components.ui.progress

import android.animation.ValueAnimator
import android.graphics.Canvas
import android.graphics.Path
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.view.animation.Interpolator
import android.view.animation.LinearInterpolator

internal class ShiftDrawable @JvmOverloads constructor(
    drawable: Drawable,
    duration: Int = DEFAULT_DURATION,
    interpolator: Interpolator? = LinearInterpolator()
) : DrawableWrapper(drawable) {

    private val animator = ValueAnimator.ofFloat(0f, 1f)
    private val visibleRect = Rect()
    private var path: Path? = null

    init {
        animator.duration = duration.toLong()
        animator.repeatCount = ValueAnimator.INFINITE
        animator.interpolator = interpolator ?: LinearInterpolator()
        animator.addUpdateListener {
            if (isVisible) {
                invalidateSelf()
            }
        }
    }

    override fun setVisible(visible: Boolean, restart: Boolean): Boolean {
        return super.setVisible(visible, restart).also { isDifferent ->
            when {
                isDifferent && visible -> animator.start()
                isDifferent && !visible -> animator.end()
            }
        }
    }

    override fun onBoundsChange(bounds: Rect) {
        super.onBoundsChange(bounds)
        updateBounds()
    }

    override fun onLevelChange(level: Int): Boolean {
        val result = super.onLevelChange(level)
        updateBounds()
        return result
    }

    override fun draw(canvas: Canvas) {
        val d = wrappedDrawable
        val fraction = animator.animatedFraction
        val width = visibleRect.width()
        val offset = (width * fraction).toInt()
        val stack = canvas.save()

        canvas.clipPath(path!!)

        // shift from right to left.
        // draw left-half part
        canvas.save()
        canvas.translate((-offset).toFloat(), 0f)
        d.draw(canvas)
        canvas.restore()

        // draw right-half part
        canvas.save()
        canvas.translate((width - offset).toFloat(), 0f)
        d.draw(canvas)
        canvas.restore()

        canvas.restoreToCount(stack)
    }

    private fun updateBounds() {
        val b = bounds
        val width = (b.width().toFloat() * level / MAX_LEVEL).toInt()
        val radius = b.height() / 2f
        visibleRect.set(b.left, b.top, b.left + width, b.height())

        // draw round to head of progressbar. I know it looks stupid, don't blame me now.
        path = Path().apply {
            addRect(
                b.left.toFloat(),
                b.top.toFloat(),
                b.left + width - radius,
                b.height().toFloat(),
                Path.Direction.CCW)
            addCircle(b.left + width - radius, radius, radius, Path.Direction.CCW)
        }
    }

    companion object {

        // align to ScaleDrawable implementation
        private const val MAX_LEVEL = 10000

        private const val DEFAULT_DURATION = 1000
    }
}
